File: src/handlers/summary.rs
use axum::{routing::get, Router, Json, extract::State};
use crate::db::AppState;
use crate::models::*;
use crate::error::AppError;

pub fn routes() -> Router<AppState> {
    Router::new().route("/summary", get(get_summary))
}

async fn get_summary(State(state): State<AppState>) -> Result<Json<Summary>, AppError> {
    let db = state.db.read().await;
    let mut total_investment = 0.0;
    let mut total_fees = 0.0;
    let mut total_tax = 0.0;
    let mut total_dividends = 0.0;
    let mut total_realized_pnl = 0.0;
    let mut cost_basis_of_holdings = 0.0;

    let assets: Vec<AssetSummary> = db.assets.iter().map(|asset| {
        let txns: Vec<&Transaction> = db.transactions.iter().filter(|t| t.asset_id == asset.id).collect();
        let buy_txns: Vec<&&Transaction> = txns.iter().filter(|t| t.txn_type == "BUY").collect();
        let sell_txns: Vec<&&Transaction> = txns.iter().filter(|t| t.txn_type == "SELL").collect();
        let buy_qty: f64 = buy_txns.iter().map(|t| t.quantity).sum();
        let sell_qty: f64 = sell_txns.iter().map(|t| t.quantity).sum();
        let holding = buy_qty - sell_qty;
        let buy_cost: f64 = buy_txns.iter().map(|t| t.price * t.quantity + t.fee + t.tax).sum();
        let sell_revenue: f64 = sell_txns.iter().map(|t| t.price * t.quantity - t.fee - t.tax).sum();
        let fees: f64 = txns.iter().map(|t| t.fee).sum();
        let tax: f64 = txns.iter().map(|t| t.tax).sum();
        let dividends: f64 = txns.iter().filter(|t| t.txn_type == "DIVIDEND").map(|t| t.quantity * t.price - t.tax).sum();
        let avg_buy_cost = if buy_qty > 0.0 { buy_cost / buy_qty } else { 0.0 };
        let holding_cost_basis = holding * avg_buy_cost;
        total_investment += buy_cost;
        total_fees += fees;
        total_tax += tax;
        total_dividends += dividends;
        cost_basis_of_holdings += holding_cost_basis;
        let realized_pnl = if holding <= 0.0001 { Some(sell_revenue - buy_cost) } else { None };
        if let Some(pnl) = realized_pnl { total_realized_pnl += pnl; }
        let tags: Vec<Tag> = db.asset_tags.iter()
            .filter(|at| at.asset_id == asset.id)
            .filter_map(|at| db.tags.iter().find(|t| t.id == at.tag_id).cloned())
            .collect();
        AssetSummary {
            id: asset.id,
            name: asset.name.clone(),
            symbol: asset.symbol.clone(),
            asset_type: asset.asset_type.clone(),
            currency: asset.currency.clone(),
            holding,
            buy_cost,
            sell_revenue,
            avg_buy_cost,
            holding_cost_basis,
            realized_pnl,
            tags,
        }
    }).collect();

    let mut type_dist = std::collections::HashMap::new();
    for a in &assets {
        *type_dist.entry(a.asset_type.clone()).or_insert(0.0) += a.holding_cost_basis;
    }

    let mut curr_dist = std::collections::HashMap::new();
    for a in &db.assets {
        let buy_cost: f64 = db.transactions.iter()
            .filter(|t| t.asset_id == a.id && t.txn_type == "BUY")
            .map(|t| t.price * t.quantity)
            .sum();
        *curr_dist.entry(a.currency.clone()).or_insert(0.0) += buy_cost;
    }

    Ok(Json(Summary {
        overview: Overview {
            total_assets: db.assets.len(),
            total_investment,
            total_realized_pnl,
            total_dividends,
            total_fees,
            total_tax,
            cost_basis_of_holdings,
        },
        type_distribution: type_dist,
        currency_distribution: curr_dist,
        assets,
    }))
}

File: src/routes/summary.ts
import type { FastifyInstance } from 'fastify';
import type { Summary, Overview, AssetSummary, Tag } from '../types';

export default async function summaryRoutes(app: FastifyInstance) {
  const db = app.db;

  // GET /api/summary - Investment summary
  app.get<{ Reply: Summary }>('/summary', { onRequest: [app.authenticate] }, async (request, reply) => {
    const userId = request.user!.sub;
    const database = db.getUserDatabase(userId);
    let totalInvestment = 0;
    let totalFees = 0;
    let totalTax = 0;
    let totalDividends = 0;
    let totalRealizedPnl = 0;
    let costBasisOfHoldings = 0;

    const assets: AssetSummary[] = database.assets.map((asset) => {
      const txns = database.transactions.filter((t) => t.asset_id === asset.id);
      const buyTxns = txns.filter((t) => t.txn_type === 'BUY');
      const sellTxns = txns.filter((t) => t.txn_type === 'SELL');
      const buyQty = buyTxns.reduce((sum, t) => sum + t.quantity, 0);
      const sellQty = sellTxns.reduce((sum, t) => sum + t.quantity, 0);
      const holding = buyQty - sellQty;
      const buyCost = buyTxns.reduce((sum, t) => sum + t.price * t.quantity + t.fee + t.tax, 0);
      const sellRevenue = sellTxns.reduce((sum, t) => sum + t.price * t.quantity - t.fee - t.tax, 0);
      const fees = txns.reduce((sum, t) => sum + t.fee, 0);
      const tax = txns.reduce((sum, t) => sum + t.tax, 0);
      const dividends = txns
        .filter((t) => t.txn_type === 'DIVIDEND')
        .reduce((sum, t) => sum + t.quantity * t.price - t.tax, 0);
      const avgBuyCost = buyQty > 0 ? buyCost / buyQty : 0;
      const holdingCostBasis = holding * avgBuyCost;

      totalInvestment += buyCost;
      totalFees += fees;
      totalTax += tax;
      totalDividends += dividends;
      costBasisOfHoldings += holdingCostBasis;

      const realizedPnl = holding <= 0.0001 ? sellRevenue - buyCost : null;
      if (realizedPnl !== null) totalRealizedPnl += realizedPnl;

      const tags: Tag[] = database.asset_tags
        .filter((at) => at.asset_id === asset.id)
        .map((at) => db.findTagById(userId, at.tag_id)?.tag)
        .filter((t): t is Tag => t !== undefined);

      return {
        id: asset.id,
        name: asset.name,
        symbol: asset.symbol,
        asset_type: asset.asset_type,
        currency: asset.currency,
        holding,
        buy_cost: buyCost,
        sell_revenue: sellRevenue,
        avg_buy_cost: avgBuyCost,
        holding_cost_basis: holdingCostBasis,
        realized_pnl: realizedPnl,
        tags,
      };
    });

    // Type distribution (by holding cost basis)
    const typeDistribution: Record<string, number> = {};
    for (const a of assets) {
      typeDistribution[a.asset_type] = (typeDistribution[a.asset_type] || 0) + a.holding_cost_basis;
    }

    // Currency distribution (by buy cost)
    const currencyDistribution: Record<string, number> = {};
    for (const a of database.assets) {
      const buyCost = database.transactions
        .filter((t) => t.asset_id === a.id && t.txn_type === 'BUY')
        .reduce((sum, t) => sum + t.price * t.quantity, 0);
      currencyDistribution[a.currency] = (currencyDistribution[a.currency] || 0) + buyCost;
    }

    const overview: Overview = {
      total_assets: database.assets.length,
      total_investment: totalInvestment,
      total_realized_pnl: totalRealizedPnl,
      total_dividends: totalDividends,
      total_fees: totalFees,
      total_tax: totalTax,
      cost_basis_of_holdings: costBasisOfHoldings,
    };

    const summary: Summary = {
      overview,
      type_distribution: typeDistribution,
      currency_distribution: currencyDistribution,
      assets,
    };

    return reply.send(summary);
  });
}
